crates/shared/src/app_data.rs

```diff
@@ -49,33 +53,259 @@ impl Validator {
+                    pre: vec![Hook {
+                        target: H160([0; 20]),
+                        call_data: vec![],
+                        gas_limit: 0,
+                    }],
+                    post: vec![
+                        Hook {
+                            target: H160([1; 20]),
+                            call_data: vec![1],
+                            gas_limit: 1
+                        },
+                        Hook {
+                            target: H160([2; 20]),
+                            call_data: vec![2, 2],
+                            gas_limit: 2,
+                        },
+                    ],
+                },
+            },
+        );
+    }
+
+    #[test]
+    fn legacy() {
+        assert_app_data!(
+            r#"
+                {
+                    "backend": {
+                        "hooks": {
+                            "pre": [
+                                {
+                                    "target": "0x0000000000000000000000000000000000000000",
+                                    "callData": "0x",
+                                    "gasLimit": "0"
+                                }
+                            ],
+                            "post": [
+                                {
+                                    "target": "0x0101010101010101010101010101010101010101",
+                                    "callData": "0x01",
+                                    "gasLimit": "1"
+                                },
+                                {
+                                    "target": "0x0202020202020202020202020202020202020202",
+                                    "callData": "0x0202",
+                                    "gasLimit": "2"
+                                }
+                            ]
+                        }
+                    }
+                }
+            "#,
+            ProtocolAppData {
+                hooks: Hooks {
+                    pre: vec![Hook {
+                        target: H160([0; 20]),
+                        call_data: vec![],
+                        gas_limit: 0,
+                    }],
+                    post: vec![
+                        Hook {
+                            target: H160([1; 20]),
+                            call_data: vec![1],
+                            gas_limit: 1
+                        },
+                        Hook {
+                            target: H160([2; 20]),
+                            call_data: vec![2, 2],
+                            gas_limit: 2,
+                        },
+                    ],
+                },
+            },
+        );
+
+        // Note that if `metadata` is specified, then the `backend` field is
+        // ignored.
+        assert_app_data!(
+            r#"
+                {
+                    "metadata": {},
+                    "backend": {
+                        "hooks": {
```

Comment:
Could this become a problem for the short term while this is live in the UI?
Would it be easy to do a db query to check for existing orders with `metadata` and `backend` keys?